src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -132,12 +134,15 @@ def resolve(self, root_reqs, check_supported_wheels):
 
             # Check if there is already an installation under the same name,
             # and set a flag for later stages to uninstall it, if needed.
-            # * There isn't, good -- no uninstalltion needed.
+            #
+            # * There is no existing installation. Nothing to uninstall.
             # * The --force-reinstall flag is set. Always reinstall.
             # * The installation is different in version or editable-ness, so
             #   we need to uninstall it to install the new distribution.
-            # * The installed version is the same as the pending distribution.
-            #   Skip this distrubiton altogether to save work.
+            # * The installed version is different from the pending distribution.
```

Comment:
Wait, this comment is misleading. We've been talking about cases where the installed version is the *same* as the pending distribution.

I hope this is a typo in the comment, not a complete misunderstading in the discussion!!!

Reply:
This is listing cases where we do install the project -- if we're installing pip 20.3beta1 over pip 20.1, we'd want that to go through. I didn't clarify this well enough.

Reply:
It's covered by the previous "The installation is different in version or editable-ness" so just remove this line, then.

But in that case, the next line "The candidate is a local wheel. Do nothing" is wrong, because it's not "a case where we do install the project" 🙁

Reply:
I'm beginning to feel like this is *hard*, and we shouldn't be rushing to decide on a behaviour like this.

It's not ideal, but why not accept the regression, ship with #8711 unfixed, as a known behaviour change in the new resolver, and open a ticket to design the correct behaviour once and for all? Yes, we'll get complaints, but people can use the legacy resolver, or use --force-reinstall, so it's not like there's no fixes, and we've already done as much as we can to tell people to ensure their workflow isn't broken when the new resolver is enabled by default...

Reply:
> why not accept the regression,

Look at the second half of https://github.com/pypa/pip/issues/8711#issuecomment-731166726.

In that, I don't like how pip is "processing" a package, then not logging anything or taking any action and exiting cleanly. This is not reasonable behavior IMO, and would pretty much definitely surprise a lot of folks.

I don't think most folks have been using the new resolver in their development CLI runs locally (by setting an env-var or setting a config option), so this will very quickly hit a lot more people than it already has.

---

This adds logging for the cases where the resolver is intentionally not doing anything, and reinstates the reinstallation in some of the other cases.

Reply:
> In that, I don't like how pip is "processing" a package, then not logging anything or taking any action and exiting cleanly. This is not reasonable behavior IMO, and would pretty much definitely surprise a lot of folks.

I'm fine with logging a message "...nothing to do" or similar, to make the fact that "Processing" doesn't mean we ended up installing anything. But IMO that's a cosmetic thing (appearances are important! but I'm personally more worried about behaviour at the moment)

> I don't think most folks have been using the new resolver in their development CLI runs locally (by setting an env-var or setting a config option), so this will very quickly hit a lot more people than it already has.

Quite probably. But if we only get bug reports post-release then we've got the information we would have liked to get earlier - it'd not our fault that people didn't test when we've worked this hard to ask them to...

But 🤷‍♂️ I'm honestly not that bothered either way. My big concern is that we're rushing decisions because we want to get 20.3 out of the door - but if we mess things up, we'll just have to fix them. So be it. Breakage now or breakage later, I'm not that fussed.

Reply:
I prefer breakage later since I feel like 20.3 is gonna have a lot of other fun things in it anyway.